File: src/main/scala/jp/scid/genomemuseum/gui/WebServiceResultTableFormat.scala
package jp.scid.genomemuseum.gui

import ca.odell.glazedlists.gui.{AdvancedTableFormat, WritableTableFormat}

import jp.scid.gui.StringSortable.ElementPropertyComparator
import jp.scid.genomemuseum.model.{SearchResult, TaskProgressModel}

/**
 * {@code SearchResult} テーブルフォーマット
 */
class WebServiceResultTableFormat extends AdvancedTableFormat[SearchResult]
    with WritableTableFormat[SearchResult] {
  import WebServiceResultTableFormat._
  
  def getColumnCount = columnNames.length
  def getColumnName(column: Int) = columnNames(column)
  def getColumnValue(e: SearchResult, column: Int) = column match {
    case 0 => e
    case 1 => e.identifier
    case 2 => e.accession
    case 3 => e.definition
    case 4 => e.length.asInstanceOf[AnyRef]
  }
  
  def getColumnClass(column: Int) = column match {
    case 0 => classOf[TaskProgressModel]
    case 4 => classOf[java.lang.Integer]
    case _ => classOf[Object]
  }
  
  def isEditable(obj: SearchResult, column: Int) = column match {
    case 0 => true
    case _ => false
  }
  
  def setColumnValue(obj: SearchResult, value: AnyRef, column: Int) = {
    obj
  } 
  
  def getColumnComparator(column: Int) = column match {
    case 0 => ElementPropertyComparator{e: SearchResult => e.identifier}
    case 1 => ElementPropertyComparator{e: SearchResult => e.identifier}
    case 2 => ElementPropertyComparator{e: SearchResult => e.accession}
    case 3 => ElementPropertyComparator{e: SearchResult => e.definition}
    case 4 => ElementPropertyComparator{e: SearchResult => Integer.valueOf(e.length)}
  }
}

object WebServiceResultTableFormat {
  private val columnNames = Array("status", "identifier",
    "accession", "definition", "length")
}

File: src/main/java/jp/scid/genomemuseum/model/WebServiceResultTableFormat.java
package jp.scid.genomemuseum.model;

import java.util.Comparator;

import ca.odell.glazedlists.gui.AdvancedTableFormat;
import ca.odell.glazedlists.gui.WritableTableFormat;

public class WebServiceResultTableFormat implements AdvancedTableFormat<NcbiEntry>, WritableTableFormat<NcbiEntry> {
    enum Column implements Comparator<NcbiEntry> {
        DOWNLOAD_ACTION(TaskProgressModel.class) {
            @Override
            NcbiEntry getColumnValue(NcbiEntry e) {
                return e;
            }
            
            @Override
            public boolean isEditable(NcbiEntry e) {
                return e.sourceUri() != null;
            }
        },
        IDENTIFIER(String.class) {
            @Override
            String getColumnValue(NcbiEntry e) {
                return e.identifier();
            }
        },
        ACCESSION(String.class) {
            @Override
            String getColumnValue(NcbiEntry e) {
                return e.accession();
            }
        },
        SEQUENCE_LENGTH(Integer.class) {
            @Override
            Integer getColumnValue(NcbiEntry e) {
                return e.sequenceLength();
            }
        },
        DEFINITION(String.class) {
            @Override
            String getColumnValue(NcbiEntry e) {
                return e.definition();
            }
        },
        TAXONOMY(String.class) {
            @Override
            String getColumnValue(NcbiEntry e) {
                return e.taxonomy();
            }
        },
        ;
        private final Class<?> dataClass;
        
        private Column(Class<?> dataClass) {
            this.dataClass = dataClass;
        }
        
        public Comparator<NcbiEntry> comparator() {
            return Comparable.class.isAssignableFrom(dataClass) ? this : null;
        }
        
        abstract Object getColumnValue(NcbiEntry e);

        @SuppressWarnings({ "unchecked", "rawtypes" })
        @Override
        public int compare(NcbiEntry o1, NcbiEntry o2) {
            Comparable val1 = (Comparable<?>) getColumnValue(o1);
            Comparable val2 = (Comparable<?>) getColumnValue(o2);
            return val1.compareTo(val2);
        }
        
        public boolean isEditable(@SuppressWarnings("unused") NcbiEntry e) {
            return false;
        }
    }
    
    @Override
    public int getColumnCount() {
        return Column.values().length;
    }

    @Override
    public String getColumnName(int column) {
        return Column.values()[column].name();
    }

    @Override
    public Object getColumnValue(NcbiEntry baseObject, int column) {
        return Column.values()[column].getColumnValue(baseObject);
    }

    @Override
    public Class<?> getColumnClass(int column) {
        return Column.values()[column].dataClass;
    }

    @Override
    public Comparator<NcbiEntry> getColumnComparator(int columnNumber) {
        return Column.values()[columnNumber].comparator();
    }

    @Override
    public boolean isEditable(NcbiEntry baseObject, int column) {
        return Column.values()[column].isEditable(baseObject);
    }

    @Override
    public NcbiEntry setColumnValue(NcbiEntry baseObject, Object editedValue, int column) {
        return baseObject;
    }
}